File: lib/lifecycle/f_using.dart
// Copyright 2023, the Freemework.ORG project authors. Please see the AUTHORS
// file for details. All rights reserved. Use of this source code is governed
// by a BSD-style license that can be found in the LICENSE file.

import "dart:async" show Future, FutureOr;

import "../execution_context/f_execution_context.dart" show FExecutionContext;
import "f_disposable.dart" show FDisposable;
import "f_initable.dart" show FInitable;

typedef FusingResourceInitializer<TResource> = FutureOr<TResource> Function(
    FExecutionContext);
typedef FusingWorker<TResource, TResult> = FutureOr<TResult> Function(
    FExecutionContext, TResource);

Future<TResult> fusing<TResource extends FDisposable, TResult>(
  final FExecutionContext executionContext,
  final FusingResourceInitializer<TResource> resourceFactory,
  final FusingWorker<TResource, TResult> worker,
) {
  Future<TResult> workerExecutor(
    final FExecutionContext workerExecutorCancellactonToken,
    TResource disposableResource,
  ) async {
    if (disposableResource is FInitable) {
      await disposableResource.init(executionContext);
    }
    try {
      final workerResult =
          worker(workerExecutorCancellactonToken, disposableResource);
      if (workerResult is Future<TResult>) {
        return await workerResult;
      } else {
        return workerResult;
      }
    } finally {
      await disposableResource.dispose();
    }
  }

  Future<TResult> workerExecutorFacade(
      final FutureOr<TResource> disposableObject) {
    if (disposableObject is Future<TResource>) {
      return disposableObject.then(
        (disposableInstance) =>
            workerExecutor(executionContext, disposableInstance),
      );
    } else {
      return workerExecutor(executionContext, disposableObject);
    }
  }

  final FutureOr<TResource> resource = resourceFactory(executionContext);
  return workerExecutorFacade(resource);
}

File: src/lifecycle/f_using.ts
import { FDisposable } from "./f_disposable";
import { FInitable, FInitableBase } from "./f_initable";
import { FExecutionContext } from "../execution_context/f_execution_context";
import { FExceptionArgument } from "../exception";


// TODO Add overloading
// return Fusing(
// 	executionContext,
// 	() => this.create(executionContext),  // <- ! initializer without executionContext
// 	(db) => worker(db)                    // <- ! worker without executionContext
// );

export namespace FUsing {
	export type ResourceInitializerWithExecutionContext<T> = (executionContext: FExecutionContext) => T | Promise<T>;
	export type ResourceInitializerWithoutExecutionContext<T> = () => T | Promise<T>;
	export type ResourceInitializer<T> = ResourceInitializerWithExecutionContext<T> | ResourceInitializerWithoutExecutionContext<T>;

	export type WorkerWithExecutionContext<TResource, TResult> = (executionContext: FExecutionContext, resource: TResource) => Result<TResult>;
	export type WorkerWithoutExecutionContext<TResource, TResult> = (resource: TResource) => Result<TResult>;
	export type Worker<TResource, TResult> = WorkerWithExecutionContext<TResource, TResult> | WorkerWithoutExecutionContext<TResource, TResult>;

	export type Result<T> = T | Promise<T>;
}
export function FUsing<TResource extends FInitable | FDisposable, TResult>(
	executionContext: FExecutionContext,
	resourceFactory: FUsing.ResourceInitializer<TResource>,
	worker: FUsing.Worker<TResource, TResult>
): Promise<TResult> {
	if (!resourceFactory || typeof resourceFactory !== "function") {
		throw new Error("Wrong argument: resourceFactory");
	}
	if (!worker) { throw new Error("Wrong argument: worker"); }

	async function workerExecutor(workerExecutorCancellationToken: FExecutionContext, disposableResource: TResource): Promise<TResult> {
		if (disposableResource instanceof FInitableBase || "init" in disposableResource) {
			await (disposableResource as FInitable).init(executionContext);
		}
		try {

			let workerResult: FUsing.Result<TResult>;
			if (worker.length === 1) {
				workerResult = (worker as FUsing.WorkerWithoutExecutionContext<TResource, TResult>)(disposableResource);
			} else if (worker.length === 2) {
				workerResult = (worker as FUsing.WorkerWithExecutionContext<TResource, TResult>)(workerExecutorCancellationToken, disposableResource);
			} else {
				throw new FExceptionArgument("Wrong worker function. Expect a function with 1 or 2 arguments.", "worker");
			}

			if (workerResult instanceof Promise) {
				return await workerResult;
			} else {
				return workerResult;
			}
		} finally {
			await disposableResource.dispose();
		}
	}

	const resource = resourceFactory(executionContext);
	if (resource instanceof Promise) {
		return (resource as Promise<TResource>)
			.then(disposableInstance => workerExecutor(executionContext, disposableInstance));
	} else {
		return workerExecutor(executionContext, resource);
	}
}
